Supported `pause()` and `resume()` for `Store`'s stream.

redboltz opened the pull request:
If `pause()` and `resume()` can be used in `Store` implementation, it
is more flexible.
For example, `mqtt-level-store` can support preserving publish order
using `pause()` and `resume()`.

https://github.com/mcollina/mqtt-level-store

Here are the documentation of `pause()` and `resume()`.

https://nodejs.org/api/stream.html#stream_readable_pause
https://nodejs.org/api/stream.html#stream_readable_resume

According to the document, `pause()` and `resume()` work well with
'data' event, but not 'readable' event. So I replaced 'readable' with 'data'.

redboltz commented:
This modification is required for https://github.com/mcollina/mqtt-level-store/pull/10.

redboltz commented:
Really? Maybe I looked over something.
I'm not 100% sure but as far as I tested, 'readable' handler is called even if the stream is paused. 
I will check again, and will write minimal and complete code that describes the situation.

redboltz commented:
> I will check again, and will write minimal and complete code that describes the situation.

I wrote it.

on 'readable' and read() version
https://wandbox.org/permlink/AvyOS2iLHUJZ0YRc

on 'data' version
https://wandbox.org/permlink/SWjdkjNE4pCt5IjB

on 'data' version works as I expected. But on 'readable' and read() version doing read() before `resume()` is called.

mcollina commented:
You should not use `pause()` and `resume()`  with `'readable'`.

Just call `.read()` to get an item, and if it's null you should wait for another `'readable'` event. There is no need to pause, because it's already paused.

`.read(0)` forces a read, which is what I think it's the problem in the current code.

redboltz commented:
I removed my update for mqtt-level-store. 

I still have a problem.

Here is mqtt-level-store branch that demonstrates the problem.
https://github.com/redboltz/mqtt-level-store/tree/readable_fail_data_ok_case

Here is the diff.
https://github.com/mcollina/mqtt-level-store/compare/master...redboltz:readable_fail_data_ok_case

I only added 2 more publishes to the test case 'should resend messages', and added console log for packets. The library code is NOT modified.

When I run the test, I got the following result.

```
Packet {
  cmd: 'publish',
  retain: false,
  qos: 1,
  dup: false,
  length: 16,
  topic: 'hello1',
  payload: <Buffer 77 6f 72 6c 64 31>,
  messageId: 54273 }
Packet {
  cmd: 'publish',
  retain: false,
  qos: 1,
  dup: false,
  length: 16,
  topic: 'hello2',
  payload: <Buffer 77 6f 72 6c 64 32>,
  messageId: 54274 }
Packet {
  cmd: 'publish',
  retain: false,
  qos: 1,
  dup: false,
  length: 16,
  topic: 'hello2',
  payload: <Buffer 77 6f 72 6c 64 32>,
  messageId: 54274 }
```

Order is not my concern in this branch because the branch is based on master. 

The problem is topic `hello2` payload `world2` published twice.
If I replaced to `on('data', ...)` version in `client.js`, this behavior is disappeared. Three packets are published as I expected.

I tried to find a way to fix the problem using `once(`readable', ...)` and `read()` approach, but I couldn't find it.

Could you tell me how to solve it using `readable` approach?

redboltz commented:
I attached the patch to replace 'readable' with 'data' on mqtt-level-store testing environment.

[readable_to_data.txt](https://github.com/mqttjs/MQTT.js/files/2320563/readable_to_data.txt)

I reported what I did on 'readable' approach.

1. Removed `outStore.read(0)`.
2. Replaced `outStore.read(1)` with `outStore.read()`.

But result is not changed.
